lib/ansible/module_utils/redfish_utils.py

```diff
@@ -0,0 +1,932 @@
+
+        for device in data[u'Members']:
+            d = device[u'@odata.id']
+            d = d.replace(self.firmware_uri, "")    # leave just device name
+            if "Installed" in d:
+                uri = self.root_uri + self.firmware_uri + d
+                # Get details for each device that is relevant
+                response = self.get_request(uri)
+                if response['ret'] is False:
+                    return response
+                result['ret'] = True
+                data = response['data']
+                firmware[data[u'Name']] = data[u'Version']
+        result["entries"] = firmware
+        return result
+
+    def get_manager_attributes(self):
+        result = {}
+        manager_attributes = {}
+        attributes_id = "Attributes"
+
+        response = self.get_request(self.root_uri + self.manager_uri + "/" + attributes_id)
+        if response['ret'] is False:
+            return response
+        result['ret'] = True
+        data = response['data']
+        for attribute in data[u'Attributes'].items():
+            manager_attributes[attribute[0]] = attribute[1]
+        result["entries"] = manager_attributes
+        return result
+
+    def get_bios_attributes(self):
+        result = {}
+        bios_attributes = {}
+        key = "Bios"
+
+        # Search for 'key' entry and extract URI from it
+        response = self.get_request(self.root_uri + self.systems_uri)
+        if response['ret'] is False:
+            return response
+        result['ret'] = True
+        data = response['data']
+        bios_uri = data[key]["@odata.id"]
+
+        response = self.get_request(self.root_uri + bios_uri)
+        if response['ret'] is False:
+            return response
+        result['ret'] = True
+        data = response['data']
+        for attribute in data[u'Attributes'].items():
+            bios_attributes[attribute[0]] = attribute[1]
+        result["entries"] = bios_attributes
+        return result
+
+    def get_bios_boot_order(self):
+        result = {}
+        boot_device_list = []
+        boot_device_details = []
+        key = "Bios"
+        bootsources = "BootSources"
+
+        # Search for 'key' entry and extract URI from it
+        response = self.get_request(self.root_uri + self.systems_uri)
+        if response['ret'] is False:
+            return response
+        result['ret'] = True
+        data = response['data']
+        bios_uri = data[key]["@odata.id"]
+
+        # Get boot mode first as it will determine what attribute to read
+        response = self.get_request(self.root_uri + bios_uri)
+        if response['ret'] is False:
+            return response
+        data = response['data']
+        boot_mode = data[u'Attributes']["BootMode"]
+        if boot_mode == "Uefi":
+            boot_seq = "UefiBootSeq"
+        else:
+            boot_seq = "BootSeq"
+
+        response = self.get_request(self.root_uri + self.systems_uri + "/" + bootsources)
```

Comment:
Concatenating this way will produce bad path in `self.systems_uri` has a trailing slash. And `BootSources` is something that I cannot find in any of the recordings or in Redfish standard, so I must assume that this is something Dell specific again.